pyasic/ssh/braiins_os.py

```diff
@@ -93,3 +103,71 @@ async def get_led_status(self):
             str: Status of the LED.
         """
         return await self.send_command("cat /sys/class/leds/'Red LED'/delay_off")
+
+    async def upgrade_firmware(self, file_location: str = None):
+        """
+        Upgrade the firmware of the BOSMiner device.
+
+        Args:
+            file_location (str): The local file path of the firmware to be uploaded. If not provided, the firmware will be downloaded from the internal server.
+
+        Returns:
+            str: Confirmation message after upgrading the firmware.
+        """
+        try:
+            logger.info("Starting firmware upgrade process.")
+
+            if file_location is None:
+                # Check for cached firmware file
+                cached_file_location = "/tmp/cached_firmware.tar.gz"
+                if os.path.exists(cached_file_location):
```

Comment:
Use [pathlib](https://docs.python.org/3/library/pathlib.html) for path handling.